added concept topic for xdcr mobile (#915)
couchbaselabs/docs-sync-gateway: modules/server-compatibility/pages/server-compatibility-xdcr-mobile.adoc

This relies on a change in how Sync Gateway tracks document revisions, transitioning from revision trees to version vectors, which enables consistent revision tracking across all participating clusters.

== Revision Tracking

Sync Gateway 4.0 fundamentally changes how it tracks document revisions to enable bi-directional XDCR between mobile clusters.

SGW 4.0 switches from revision trees to version vectors for implementing and managing bi-directional XDCR deployments.

=== Revision Trees

Sync Gateway versions prior to 4.0 tracked document revisions using revision trees with Multi-Version Concurrency Control (MVCC).

This approach created a tree structure for each document where branches represented conflicting updates and nodes contained revision IDs.
While effective for mobile replication protocols, revision trees maintained cluster-specific metadata that prevented meaningful bi-directional replication with XDCR.
Each cluster maintained its own revision tree, making it impossible to preserve revision history when documents replicated between clusters.

=== Version Vectors

Version vectors provide a proven mechanism for tracking causality in distributed systems.

> A version vector consists of logical clocks maintained by each component in the distributed environment.

Reviewer: ```suggestion
A version vector consists of hybrid logical clocks maintained by each component in the distributed environment.
```

https://martinfowler.com/articles/patterns-of-distributed-systems/hybrid-clock.html

hybrid logical clocks incorporate time into a simple logical clock

> Sync Gateway, XDCR, and Couchbase Lite all operate as independent components.

Reviewer: There's a subtle thing here, but _each instance_ of Couchbase Lite is its own component in the version vector.
SG and XDCR both interact with the component associated with the cluster-specific Couchbase Server bucket.

It's less about each architectural component (CB Product), and more about logical "actors" capable of updating the document.
I think the "many CBL components" thing is an important distinction to have, though I'm less bothered about the SG/XDCR/Bucket part.

Author: Thanks for the clarification! I've updated the section to address your feedback.

Each component increments its own count when making changes, creating a vector that captures the complete update history across all clusters.

This architectural change enables:

* Unified revision tracking that both XDCR and Sync Gateway understand
* Consistent revision identification across all clusters
* Preservation of causality relationships between document versions
* Automatic conflict detection based on vector comparison

The version vector metadata travels with documents during replication, ensuring that revision history remains intact regardless of which cluster processes the document.

==== Conflict Resolution with Version Vectors

The shift to version vectors changes the default conflict resolution strategy from revision trees' Most-Write-Wins to Last-Write-Wins.

> Last-Write-Wins uses timestamp comparison to resolve conflicts deterministically across all clusters.

Reviewer: ```suggestion
Last-Write-Wins uses hybrid logical clock timestamp comparison to resolve conflicts deterministically across all clusters.
```

We have prior-art for using `hybrid logical clock timestamp` in XDCR docs:

https://docs.couchbase.com/server/current/learn/clusters-and-availability/xdcr-conflict-resolution.html#timestamp-based-conflict-resolution

image::ROOT:[]

// IMAGE

=== Metadata Filtering

XDCR automatically filters cluster-specific metadata to prevent corruption of local state.

System documents with the `_sync:` prefix remain local to each cluster -- these include sequence trackers, user definitions, and role assignments.
Document-level sync metadata in extended attributes gets selectively replicated, with cluster-local information like sequence numbers filtered while global metadata like version vectors passes through.
This filtering happens transparently without requiring manual configuration.

== Configuration Requirements

Enabling bi-directional XDCR requires specific configuration at both the bucket and replication levels.

=== Bucket Configuration

Enable cross-cluster versioning on all participating buckets to support version vectors.

Apply the `enableCrossClusterVersioning=true` configuration to buckets on both source and target clusters before establishing replication.

=== Sync Gateway Configuration

Configure Sync Gateway nodes with shared bucket access and import processing enabled.

Add these settings to the database configuration:
[source,json]
----
{
  "databases": {
    "db": {

>       "enable_shared_bucket_access": true,

Reviewer: `enable_shared_bucket_access` is being removed as an option in 4.0 since we're enforcing it to be set

`import_docs` also defaults to true (in Enterprise Edition) - so there's no specific config required on SG side other than running 4.0 to make this work - just the XDCR setting.

Author: I've updated the section for better clarity.

      "import_docs": true
    }
  }
}
----

Bi-directional XDCR requires both settings.
Shared bucket access allows Sync Gateway to coexist with direct SDK writes, while import processing handles replicated documents.

=== Two Active-Active Clusters

The configuration involves two clusters with bi-directional replication between them.

Both clusters run Sync Gateway in active mode, processing reads and writes from mobile clients.
Load balancers distribute traffic based on geographic proximity or availability.
This topology provides the foundation for zero-downtime disaster recovery and global data distribution.
Testing and validation have focused primarily on two-cluster deployments.

=== Checkpoint Management

Clients maintain independent checkpoints for each cluster they connect to.

> When a client connects to a cluster for the first time, it begins replication from sequence zero, downloading all accessible documents.

Reviewer: Important distinction:

Clients don't (re)download _documents_ when checkpoints are reset, they get change notifications, which _can_ result in a document being downloaded (but only if missing)

This is important since we're not re-replicating data unnecessarily here, it's just _metadata_ about changes.

```suggestion
When a client connects to a cluster for the first time, it begins replication from sequence zero, checking which documents are missing and need downloading.
```

=== Sync Gateway Configuration

Sync Gateway 4.0 requires no additional configuration for bi-directional XDCR support.

Sync Gateway 4.0 enforces shared bucket access by default, enabling coexistence with direct SDK writes and XDCR replication.

> Couchbase Server Enterprise Edition enables document import processing by default, automatically handling replicated documents from XDCR.

Reviewer: ```suggestion
Sync Gateway Enterprise Edition enables document import processing by default, automatically handling replicated documents from XDCR.
```
